packages/connect-popup/src/index.tsx

```diff
@@ -218,70 +141,215 @@ const handleMessage = (
+        if (!payload.systemInfo) {
+            payload.systemInfo = getSystemInfo(config.supportedBrowsers);
+        }
+
+        const isBrowserSupported = await view.initBrowserView(payload.systemInfo);
+        log.debug('browser supported: ', isBrowserSupported);
+        if (!isBrowserSupported) {
+            return;
+        }
+
         // render react view
-        await renderConnectUI();
-        reactEventBus.dispatch({ type: 'waiting-for-iframe-init' });
-        await initMessageChannel(payload, handleMessage);
-        reactEventBus.dispatch({ type: 'waiting-for-iframe-handshake' });
+        renderConnectUIPromise = renderConnectUI();
+        await renderConnectUIPromise;
+        log.debug('connect-ui rendered');
+
+        addWindowEventListener('message', handleMessageInIframeMode, false);
+        await initCoreInIframe(payload);
     } catch (error) {
         postMessageToParent(createPopupMessage(POPUP.ERROR, { error: error.message }));
     }
 };
 
+const initCoreInIframe = async (payload: PopupInit['payload']) => {
+    reactEventBus.dispatch({ type: 'loading', message: 'waiting for iframe init' });
+    await initMessageChannelWithIframe(payload, handleMessageInIframeMode);
+    // done, popup is ready to handle incoming messages, waiting for handshake from iframe
+};
+
 // handle POPUP.HANDSHAKE message from iframe
-const handshake = (handshake: PopupHandshake) => {
+const handshake = (handshake: PopupHandshake, origin: string) => {
     const { payload } = handshake;
+    log.debug('handshake with origin: ', origin, 'payload: ', payload);
+
     if (!payload) return;
 
     clearTimeout(handshakeTimeout);
 
-    // use trusted settings from iframe
-    setState({ settings: payload.settings });
+    // when this message comes from iframe, settings is already validated.
+    const trustedSettings = payload.settings;
+    setState({ settings: trustedSettings });
 
-    reactEventBus.dispatch(handshake);
-
-    if (isPhishingDomain(payload.settings.origin || '')) {
+    if (isPhishingDomain(trustedSettings.origin || '')) {
         reactEventBus.dispatch({ type: 'phishing-domain' });
     }
+
+    reactEventBus.dispatch(handshake);
+
     log.debug('handshake done');
 };
 
+// register onLoad event
 const onLoad = () => {
     postMessageToParent(createPopupMessage(POPUP.LOADED));
+    handshakeTimeout = setTimeout(() => {
+        fail({
+            type: 'error',
+            detail: 'handshake-timeout',
+        });
+    }, 90 * 1000);
+};
+
+/**
+ * keep track of all window listeners
+ */
+const windowEventListeners: { type: keyof WindowEventMap; listener: (args: any) => void }[] = [];
+const addWindowEventListener = <K extends keyof WindowEventMap>(
+    type: K,
+    listener: (this: Window, ev: WindowEventMap[K]) => any,
+    options?: boolean | AddEventListenerOptions,
+) => {
+    windowEventListeners.push({ type, listener });
+    window.addEventListener(type, listener, options);
+};
+
+let disposed = false;
```

Comment:
this newly introduced `disposed` global var is not very beatufil. I am open to a better solution, this is an easy and working on. What I need it for: 

1. message A arrives
2. async action is triggered
3. popup turns into error state
4. async action is finished, possibly doing something we don't want. 

I am using disposed here to check after every async action whether execution should continue or not.

Reply:
Ok, I think I understand, but let's say there is some error and dispose is set to `true`, when is it going to previous state `false`? When popup closes and opens again?
It probably makes sense, since popup is basically one time use.

Reply:
yes, never, this is "set one time only". meaning that popup died. nothing should happen anymore.